### ISSUE TITLE
Incorrect and unstable MD5 hashing.

### ISSUE THREAD

**Issue by ImagicTheCat** [love2d/love#1932]
Hi,

main.lua:
```lua
local data = "test"
local i = 0
local hash
function love.update(dt)
  i = i+1
  local new_hash = love.data.hash("md5", data)
  assert((not hash) or (hash == new_hash))
  hash = new_hash
  print(i, data, love.data.encode("string", "hex", hash))
end
```

Will throw an error for `md5`, but will be fine for other hash functions.

I suspected that there was some corruption or uninitialized memory, so I printed the content of the prepared data.

```diff
diff --git a/src/modules/data/HashFunction.cpp b/src/modules/data/HashFunction.cpp
index 643d180b..bd2e9c47 100644
--- a/src/modules/data/HashFunction.cpp
+++ b/src/modules/data/HashFunction.cpp
@@ -95,6 +95,10 @@ class MD5 : public HashFunction
 		*((uint64*) &padded[paddedLength]) = length * 8;
 		paddedLength += 8;

+    for(int i = 0; i < paddedLength; i++)
+      printf("%d ", (int)padded[i]);
+    printf("%s", "\n");
+
 		for (uint64 i = 0; i < paddedLength; i += 64)
 		{
 			uint32 *chunk = (uint32*) &padded[i];
```

C++ is not part of my methodology and I don't understand clearly what the code is doing, but with that code the hash is correct and stable, thus I suspect that there are some undefined behaviors going on. Maybe a case of strict aliasing violation.

It happens with LÖVE 11.4: the binary provided by Arch Linux `love 11.4-1` and a manually compiled version from tag `11.4`.

**Comment by ImagicTheCat** [comment]
(After a bit of [C homework](https://en.cppreference.com/w/c/language/object))

As I assumed, the hash implementations are violating strict aliasing. It would be better to rely on a good hashing dependency, easy to integrate in LÖVE. But I suppose it wasn't, maybe still isn't, easy to find one.

I made a fix for the MD5 function. It assumes that `paddedLength` is a multiple of 4, which I didn't verified yet.

```diff
diff --git a/src/modules/data/HashFunction.cpp b/src/modules/data/HashFunction.cpp
index 34d9e008..307e1dc9 100644
--- a/src/modules/data/HashFunction.cpp
+++ b/src/modules/data/HashFunction.cpp
@@ -86,18 +86,19 @@ class MD5 : public HashFunction
 		if (paddedLength % 64 > 56)
 			paddedLength += 120 - paddedLength % 64;

-		uint8 *padded = new uint8[paddedLength + 8];
+		uint32 *padded = new uint32[(paddedLength + 8)/4];
 		memcpy(padded, input, length);
-		memset(padded + length, 0, paddedLength - length);
-		padded[length] = 0x80;
+		memset(((uint8*)padded) + length, 0, paddedLength - length);
+		*(((uint8*)padded) + length) = 0x80;

 		//Now we need the length in bits
-		*((uint64*) &padded[paddedLength]) = length * 8;
+		uint64 bit_length = length * 8;
+		memcpy(((uint8*)padded) + paddedLength, &bit_length, sizeof(bit_length));
 		paddedLength += 8;

 		for (uint64 i = 0; i < paddedLength; i += 64)
 		{
-			uint32 *chunk = (uint32*) &padded[i];
+			uint32 *chunk = &padded[i];

 			uint32 A = a0;
 			uint32 B = b0;
```

If using a dependency is not possible, I could do a pull request to apply the fix to all functions.

**Comment by slime73** [comment]
> I could do a pull request to apply the fix to all functions.

That'd be appreciated, thanks!

I haven't looked thoroughly enough to be 100% positive but I think your MD5 change might be good after changing the allocation to align its size up to the next multiple of 4 bytes (so it won't round down during the integer division) - something like this:
```cpp
// Round up instead of down when dividing by 4, and add an extra 64 bits at the end.
uint32 *padded = new uint32[(paddedLength + 8 + 3) / 4];
```

**Comment by ImagicTheCat** [comment]
The message is padded to be a multiple of 64 bytes, thus will be multiple of 4, so nothing has to be done. Obvious if we consider that most hash functions work on fixed blocks, in this case *16 x 32bit words*.

I will review all functions and make a pull request.

**Comment by slime73** [comment]
#1935 has been merged (in [63f0af8](https://github.com/love2d/love/commit/63f0af8971dd6feb25d94c9e50c91c025a694bfa)) which fixes this.
